server/src/internalClusterTest/java/org/elasticsearch/index/mapper/IgnoreFieldIT.java

```diff
@@ -0,0 +1,144 @@
+import static org.elasticsearch.search.aggregations.AggregationBuilders.avg;
+import static org.elasticsearch.search.aggregations.AggregationBuilders.terms;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+import static org.hamcrest.Matchers.hasSize;
+
+@SuppressWarnings("resource")
+public class IgnoreFieldIT extends ESIntegTestCase {
+
+    public static final String NUMERIC_FIELD_NAME = "numeric_field";
+    public static final String DATE_FIELD_NAME = "date_field";
+    public static final String TEST_INDEX = "test-index";
+    public static final String CORRECT_FIELD_TYPE_DOC_ID = "1";
+    public static final String WRONG_FIELD_TYPE_DOC_ID = "2";
+
+    @Before
+    public void createTestIndex() throws Exception {
+        XContentBuilder mapping = XContentFactory.jsonBuilder()
+            .startObject()
+            .startObject("_doc")
+            .startObject("properties")
+            .startObject(NUMERIC_FIELD_NAME)
+            .field("type", "long")
+            .field("ignore_malformed", true)
+            .endObject()
+            .startObject(DATE_FIELD_NAME)
+            .field("type", "date")
+            .field("ignore_malformed", true)
+            .endObject()
+            .endObject()
+            .endObject()
+            .endObject();
+        assertAcked(indicesAdmin().prepareCreate(TEST_INDEX).setMapping(mapping));
+        indexTestDoc(NUMERIC_FIELD_NAME, CORRECT_FIELD_TYPE_DOC_ID, "42");
+        indexTestDoc(NUMERIC_FIELD_NAME, WRONG_FIELD_TYPE_DOC_ID, "forty-two");
+    }
+
+    public void testIgnoreFieldFetch() {
+        SearchResponse searchResponse = prepareSearch().setQuery(new IdsQueryBuilder().addIds(CORRECT_FIELD_TYPE_DOC_ID))
+            .addFetchField(NUMERIC_FIELD_NAME)
+            .get();
+        assertHitCount(searchResponse, 1);
+        SearchHit hit = searchResponse.getHits().getAt(0);
+        DocumentField numericField = hit.field(NUMERIC_FIELD_NAME);
+        assertNotNull(numericField);
+        assertEquals(42, (long) numericField.getValue());
+        DocumentField ignoredField = hit.field(IgnoredFieldMapper.NAME);
+        assertNull(ignoredField);
+
+        searchResponse = prepareSearch().setQuery(new IdsQueryBuilder().addIds(WRONG_FIELD_TYPE_DOC_ID))
+            .addFetchField(NUMERIC_FIELD_NAME)
+            .get();
+        assertHitCount(searchResponse, 1);
+        hit = searchResponse.getHits().getAt(0);
+        numericField = hit.field(NUMERIC_FIELD_NAME);
+        assertNotNull(numericField);
+        assertEquals("forty-two", numericField.getIgnoredValues().get(0));
+        ignoredField = hit.field(IgnoredFieldMapper.NAME);
+        assertNotNull(ignoredField);
+        assertEquals(NUMERIC_FIELD_NAME, ignoredField.getValue());
+    }
+
+    public void testExistsQuery() {
```

Comment:
This should probably be added as a unit test to IgnoredFieldTypeTests? Same probably for the other two tests you have? Can we also have a unit test that verifies the versioning logic?